### PR TITLE
feat: support keycloak clustered deployment using JDBC_PING for quarkus distribution

### LINKED ISSUE #199: Quarkus versions not deployable (keycloak > 16)
The quarkus versions of Keycloak (versions > 16) are not deployable with cdk-keycloak.

The setup of the stack works correctly, except the startup of the Keycloak service in fargate:

The issues are related to:
- Database connection
  - Database connection to singleDbInstance configurations are not working: `quarkus.datasource.jdbc.driver is set to 'com.mysql.cj.jdbc.MysqlXADataSource' but it is build time fixed to 'org.h2.jdbcx.JdbcDataSource'`
  - In aurora setup the same: Connection url is not supported: mysql.
  - Issue comes with the hardcoded "--optimized" startup parameter, that only has support in the default  
- JGroups causes failing startup:
  - `Error while trying to create a channel using the specified configuration file: default-configs/default-jgroups-ec2.xml` 
  -  and ` unable to load protocol org.jgroups.aws.s3.NATIVE_S3_PING (either with relative - org.jgroups.aws.s3.NATIVE_S3_PING - or absolute - org.jgroups.protocols.org.jgroups.aws.s3.NATIVE_S3_PING - class name)`

Tested with  keycloakVersion settings  `kc.KeycloakVersion.V19_0_3`,  `kc.KeycloakVersion.V20_0_3`,  `kc.KeycloakVersion.V21_0_1` and `kc.KeycloakVersion.of('21.1.1')`.

Or is there a suggested workaround / different approach to use a custom image? Then this should be at least documented in the readme.

### PR DESCRIPTION
**Pull request by hpp222**
Fixes #199

- Quarkus version is deployable by using JDBC_PING discovery protocol
- keycloak version 22.0.4 tested in 5 regions: us-east-1, us-west-2, ap-southeast-1, cn-north-1, cn-northwest-1

### CODE REVIEW

File: src/keycloak.ts [hunk 1]
```diff
@@ -575,19 +592,30 @@ export class Database extends Construct {
...
+      writer: rds.ClusterInstance.provisioned('Writer', {
+        instanceType: instanceProps.instanceType,
+        isFromLegacyInstanceProps: instanceProps.isFromLegacyInstanceProps,
+      }),
+      readers: [
+        rds.ClusterInstance.provisioned('Reader', {
+          instanceType: instanceProps.instanceType,
+          isFromLegacyInstanceProps: instanceProps.isFromLegacyInstanceProps,
+        }),
+      ],
```
Comment (wchaws) [inline src/keycloak.ts:618]: Is this separate read/write instances?  Is there any reason why we need reader/writer at here?

File: src/keycloak.ts [hunk 2]
```diff
@@ -528,19 +534,30 @@ export class Database extends Construct {
...
+      parameterGroup: rds.ParameterGroup.fromParameterGroupName(this, 'ParameterGroup', 'default.aurora-mysql8.0'),
```
Comment (wchaws) [inline src/keycloak.ts:560]: Is there any reason why we need reader/writer at here?

File: entryPoint.txt
Comment (wchaws) [inline entryPoint.txt:1]: I think you could directly embed this file into a typescript variable

File: src/keycloak.ts [hunk 3]
```diff
@@ -719,11 +747,10 @@ export class ContainerService extends Construct {
   constructor(scope: Construct, id: string, props: ContainerServiceProps) {
     super(scope, id);
 
-    const region = cdk.Stack.of(this).region;
     let containerPort = 8443;
     let protocol = elbv2.ApplicationProtocol.HTTPS;
-    let command = undefined;
-    let s3PingBucket: s3.Bucket | undefined = undefined;
+    let entryPoint = fs.readFileSync(path.join(__dirname, '../entryPoint.txt'), 'utf8').split(',');
```
Comment (wchaws) [inline src/keycloak.ts:752]: Please embed the config file directly into a typescript variable

File: src/keycloak.ts [hunk 4]
```diff
@@ -774,6 +794,9 @@ export class ContainerService extends Construct {
         KC_HOSTNAME: props.hostname!,
         KC_HOSTNAME_STRICT_BACKCHANNEL: 'true',
         KC_PROXY: 'edge',
+        INITIALIZE_SQL: 'CREATE TABLE IF NOT EXISTS JGROUPSPING (own_addr varchar(200) NOT NULL, cluster_name varchar(200) NOT NULL, ping_data VARBINARY(255), constraint PK_JGROUPSPING PRIMARY KEY (own_addr, cluster_name));',
```
Comment (wchaws) [inline src/keycloak.ts:797]: I don't think you need to add this ENV if DB engine is mysql. You could check this https://github.com/aws-samples/keycloak-on-aws/issues/95

**Review by wchaws — CHANGES_REQUESTED** [review]
Please add some description about this PR which aws regions did you test.

File: src/keycloak.ts [hunk 5]
```diff
@@ -108,7 +114,7 @@ interface dockerImageMap {
 
 const KEYCLOAK_DOCKER_IMAGE_URI_MAP: dockerImageMap = {
   'aws': 'quay.io/keycloak/keycloak:',
-  'aws-cn': '048912060910.dkr.ecr.cn-northwest-1.amazonaws.com.cn/dockerhub/jboss/keycloak:',
+  'aws-cn': '048912060910.dkr.ecr.cn-northwest-1.amazonaws.com.cn/quay/keycloak/keycloak:',
```
Comment (wchaws) [inline src/keycloak.ts:117]: Please make sure mirror existing

File: src/keycloak.ts (hunk 2 above)
Comment (hpp222) [reply src/keycloak.ts:560]: aws-cdk-lib.aws_rds.DatabaseClusterProps#instanceProps is deprecated, we should use writer and readers instead. See https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_rds-readme.html#migrating-from-instanceprops

File: src/keycloak.ts (hunk 4 above)
Comment (hpp222) [reply src/keycloak.ts:797]: Thanks. Tests show cache-ispn-jdbc-ping.xml works without initialize_sql property for mysql engine.

File: src/keycloak.ts (hunk 5 above)
Comment (hpp222) [reply src/keycloak.ts:115]: I'm requesting syncing the latest versions to NWCD ECR, see [here](https://github.com/nwcdlabs/container-mirror/pull/215/files) for more information.

**Review by wchaws — APPROVED** [review]
LGTM

**Pull request merged**